### DISCUSSION TITLE
Customization: Problems with discrete Variable

### DISCUSSION THREAD

**Discussion by dymovar96** [anyoptimization/pymoo#239] [Q&A] [answered]
Hello,

I am aiming to solve single-objective and multi-objective optimization problems with different variables' types. I found out how to model mixed variable problems and have successfully tested continuous and integer variables problems with GA. But I have not understood how to declare a discrete variable, that has N non-equispaced values. 

For example, there is a cantilevered_beam.py in the problem folder. 

    def __init__(self):
        super().__init__(n_var=4, n_obj=1, n_constr=2, type_var=np.double)
        self.xl = np.array([2, 0.1, 0.1, 3.0])
        self.xu = np.array([12.0, 1.0, 2.0, 7.0])
        self.h1 = np.array([0.1, 0.25, 0.35, 0.5, 0.65, 0.75, 0.9, 1.0])

It seems that h1 is defined as a non-equispaced discrete variable.  For mixed variables, different evolutionary operators must be applied to different types of variables. In pymoo this is supported by instantiating operators with a mask. 

For example for integers and continues:

    mask = ["int", "real","real", "real"]

    sampling = MixedVariableSampling(mask, {
        "real": get_sampling("real_random"),
        "int": get_sampling("int_random")

How should I modify the mask and operators in order to handle non-equispaced discrete variables and solve this problem? 
Best regards,
Artem

**Comment by blankjul** [comment, marked as answer]
An easy fix for that is to map the variables in the `_evaluate` function.

Let us say you define the discrete variables [1, 2, 3] and then map 1->10, 2->100, and 3->1000.
This way the algorithm just handles ordered integers, but your problem definition uses them in a non-equidistant manner.

**Reply by dymovar96** [reply]
Thank you! So simple, it works!
